src/main/java/org/jvnet/hudson/reactor/Reactor.java

```diff
@@ -173,7 +173,11 @@ private synchronized Node milestone(final Milestone m) {
         if (n==null) {
             milestones.put(m,n=new Node(new Runnable() {
                 public void run() {
-                    listener.onAttained(m);
+                    try {
+                        listener.onAttained(m);
+                    } catch(Throwable x) {
+                        throw new TunnelException(x);
```

Comment:
It suppresses `Error`s as well. Generally it is not a recommended practice in Java, but Jenkins uses this approach in some cases. Maybe makes sense to rethrow errors

Reply:
I reproduced what was done for non-milestone tasks.
If we should rethrow errors, then it should be done for both cases, and probably in Reactor.run method to prevent deadlock (especially as it was an AssertionError that made the deadlock occur in the first case).